File: coding/src/com/algos/core/li15_dp/revised/Strings_Lcs_LongestCommonSubstring.java
package  com.algos.core.li15_dp.revised;

public class Strings_Lcs_LongestCommonSubstring {

	public static void main(String[] args) {
		String str1 = "abcdaf";
		String str2 = "zbcdf";
		System.out.println(longestCommonSubstring(str1, str2));
	}

	/*		
	 * Idea: If char1 in str1 is equal to char2 in str2, check for remaining str1 and str2 excluding char1 and char2.
	 * 		
	 *		i\j	""	a	b	c	d	a	f
	 * 		""	0	0	0	0	0	0	0
	 * 		z	0	0	0	0	0	0	0	
	 * 		b	0	0	1	0	0	0	0
	 * 		c	0	0	0	2	0	0	0
	 * 		d	0	0	0	0	3	0	0
	 * 		f	0	0	0	0	0	0	1
	 * */
	// Time: O(m*n)
    // Space: O(m*n)
	private static int longestCommonSubstring(String str1, String str2) {
		int m = str1.length(), n = str2.length();

		int[][] mem = new int[m+1][n+1];
		int max = 0;
		for(int i=1; i<m; i++) {
			for(int j=1; j<n; j++) {
				if(str1.charAt(i-1) == str2.charAt(j-1)){
					mem[i][j] = 1+mem[i-1][j-1];
					
					if(mem[i][j] > max) {
						max = mem[i][j];
					}
				}
			}
		}
		return max;
	}

}

File: dsa/patterns/15_dp/python/strings_lcs_longest_common_substring.py
# DP / Strings
#
# Problem: Longest Common Substring
# Idea:
#   - If str1[i-1] == str2[j-1], extend the substring by 1 from previous dp[i-1][j-1]
#   - Keep track of the maximum length
# Time: O(m*n)
# Space: O(m*n)

def longestCommonSubstring(str1, str2):
    m, n = len(str1), len(str2)
    mem = [[0] * (n + 1) for _ in range(m + 1)]
    max_len = 0

    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if str1[i - 1] == str2[j - 1]:
                mem[i][j] = 1 + mem[i - 1][j - 1]
                max_len = max(max_len, mem[i][j])

    return max_len


# -------- Test --------
str1 = "abcdaf"
str2 = "zbcdf"

print(longestCommonSubstring(str1, str2))  # Expected: 4 ("bcd")
